sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,571 @@
+| __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_GEN | "pi_ext_spirv64_gen_linking" | Linking of 64-bit programs that were AOT compiled for GPU device is supported|
+| __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_FPGA | "pi_ext_spirv64_fpga_linking" | Linking of 64-bit programs that were AOT compiled for FPGA device is supported|
+
+To link several device images together `piProgramLink` API will be used.
+Depending on concrete plugin implementation and set of device image formats that
+can be linked at run-time, `piProgramLink` API may receive programs made from
+device images in different formats as inputs (including SPIR-V and native code).
+
+##### Support of runtime linking in backends
+
+- The initial implementation will support dynamic linking of device code in SPIR-V
+  format on OpenCL backend:
+  - OpenCL plugin will use the existing OpenCL `clLinkProgram()` API to online
+  link the SPIR-V modules together.
+  - A new Level Zero API to online link programs on SPIR-V level is required for
+  better performance.
+  - While there is no Level Zero API to link programs on SPIR-V level, existing
+  `zeModuleDynamicLink()` can be used as fallback.
+
+- In order to support dynamic linking of AOT compiled device code the
+  following should be implemented on backends site:
+  - AOT compilers must allow to compile SPIR-V modules with unresolved symbols
+  and produce device code in format that can be linked in run time and allows
+  to reduce JIT overhead
+  - OpenCL program binary type CL_PROGRAM_BINARY_TYPE_[COMPILED_OBJECT/LIBRARY]
+  should have native code format or any other format that can be emitted by AOT
+  compiler and allows to reduce JIT overhead
+
+#### Device images collection and linking
+
+Device images collection and linking of programs is performed by DPC++ Runtime
+class named `ProgramManager`.
+
+When the program manager gets a request to create a program object using device
+image, it examines its list of imported symbols and finds device images which
+export those symbols, then program manager creates programs for each required
+device image and links them all together.
+
+All needed device images are found by iterating through all available OS modules
+without predefined order and searching for first unresolved symbol in list of
+imports of target device image. During search device image format is taken
+into account, i.e. only device images that have the same format as target device
+image will be considered as suitable ones. Once suitable device image that
+contains first symbol is met, remaining exported symbols are checked in found
+image. If they match some imported symbols then these matched symbols will be
+marked as resolved. The procedure repeats until all imported symbols are marked
+as resolved. In case all available device images are viewed, but some imported
+symbols remain unresolved, exception will be thrown.
```

Comment:
What exception will be thrown?  If we want to use one of the existing exception codes, I think it should either be `errc::runtime` or `errc::build`.  (See [Table 136](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:exception.class) in the SYCL 2020 spec.)  I think I have a slight preference for `errc::runtime`, so that we reserve `errc::build` for exceptions that happen when the application specifically asks to online compile or link a kernel bundle.

Reply:
`errc::build` seems more suitable for me, since usually OpenCL build APIs emit build error when incoming program has unresolved symbols. In addition, in regular c++ errors about unresolved symbols are usually emitted during build time. However this is not a strong preference, if we want to use `errc::runtime`, that is ok for me.
@s-kanaev , @romanovvlad , any opinion?

Reply:
I was just looking at the DPC++ code in `program_manager`.  It looks like we currently throw `compile_program_error` if there is a JIT-time compilation error for a kernel.  This is most similar to the SYCL 2020 `errc::build` exception code, so I presume that is what we will throw in the future.  Therefore, it probably does make sense to throw `errc::build` if there are unresolved symbols.